test/browser/components.test.js

```diff
@@ -1573,4 +1573,319 @@ describe('Components', () => {
+			constructor(props, context) {
+				super(props, context);
+				child = this;
+				this.state = { tagName: 'p' };
+			}
+			render() {
+				return h(this.state.tagName, {}, 'Hello');
+			}
+		}
+
+		class Parent1 extends Component {
+			render() {
+				parent1 = this;
+				return this.props.children;
+			}
+		}
+
+		function Sibling(props) {
+			sibling = this;
+			return <p />;
+		}
+
+		function Null() {
+			nullInst = this;
+			return null;
+		}
+
+		render((
+			<Parent1>
+				<Null />
+				<Child />
+				<Sibling />
+			</Parent1>
+		), scratch);
+
+		expect(scratch.innerHTML).to.equal('<p>Hello</p><p></p>');
+		expect(nullInst.base).to.equal(undefined);
+		expect(child.base).to.equalNode(scratch.firstChild);
+		expect(sibling.base).to.equalNode(scratch.lastChild);
+		expect(parent1.base).to.equalNode(child.base);
+
+		child.setState({ tagName: 'span' });
+		rerender();
+
+		expect(scratch.innerHTML).to.equal('<span>Hello</span><p></p>');
+		expect(nullInst.base).to.equal(undefined);
+		expect(child.base).to.equalNode(scratch.firstChild);
+		expect(sibling.base).to.equalNode(scratch.lastChild);
+		expect(parent1.base).to.equalNode(child.base);
+	});
+
+	it('should not update parent c.base if child component changes DOM nodes and a parent is not first child component', () => {
+		let parentDom1;
+		let parent1;
+		let parent2;
+		let child;
+		let sibling;
+
+		class Child extends Component {
+			constructor(props, context) {
+				super(props, context);
+				child = this;
+				this.state = { tagName: 'p' };
+			}
+			render() {
+				return h(this.state.tagName, {}, 'Hello');
+			}
+		}
+
+		class Parent1 extends Component {
+			render() {
+				parent1 = this;
+				return this.props.children;
+			}
+		}
+
+		function Parent2(props) {
+			parent2 = this;
+			return props.children;
+		}
+
+		function ParentWithDom(props) {
+			parentDom1 = this;
+			return <div>{props.children}</div>;
+		}
+
+		function Sibling(props) {
+			sibling = this;
+			return <p />;
+		}
+
+		render((
+			<ParentWithDom>
+				<Parent1>
+					<Sibling />
+					<Parent2>
+						<Child />
+					</Parent2>
+				</Parent1>
+			</ParentWithDom>
+		), scratch);
+
+		expect(scratch.innerHTML).to.equal('<div><p></p><p>Hello</p></div>');
+		expect(child.base).to.equalNode(scratch.firstChild.lastChild);
+		expect(parent2.base).to.equalNode(child.base);
+		expect(sibling.base).to.equalNode(scratch.firstChild.firstChild);
+		expect(parent1.base).to.equalNode(sibling.base);
+		expect(parentDom1.base).to.equalNode(scratch.firstChild);
+
+		child.setState({ tagName: 'span' });
+		rerender();
+
+		expect(scratch.innerHTML).to.equal('<div><p></p><span>Hello</span></div>');
+		expect(child.base).to.equalNode(scratch.firstChild.lastChild);
+		expect(parent2.base).to.equalNode(child.base);
+		expect(sibling.base).to.equalNode(scratch.firstChild.firstChild);
+		expect(parent1.base).to.equalNode(sibling.base);
+		expect(parentDom1.base).to.equalNode(scratch.firstChild);
+	});
```

Comment:
I think we're missing one test scenario where the first child is non-null and unmounted on the second render. That seems to occur frequently on the apps I've worked on:

```jsx
// First render
<div>
  <Component>
    <Child />
    <Sibling />
  </Component>
</div>

// Second render
<div>
  <Component>
    {null}
    <Sibling />
  </Component>
</div>
```

Reply:
Good call! Will add